system/Database/MySQLi/Connection.php

```diff
@@ -444,71 +446,76 @@ public function _fieldData(string $table): array
 	/**
 	 * Returns an array of objects with index data
 	 *
-	 * @param	string	$table
-	 * @return	\stdClass[]
+	 * @param string $table
+	 *
+	 * @return \stdClass[]
 	 * @throws DatabaseException
 	 * @throws \LogicException
 	 */
 	public function _indexData(string $table): array
 	{
-		$table = $this->protectIdentifiers($table, TRUE, NULL, FALSE);
+		$table = $this->protectIdentifiers($table, true, null, false);
 
-		if (($query = $this->query('SHOW CREATE TABLE ' . $table)) === FALSE)
+		if (($query = $this->query('SHOW INDEX FROM ' . $table)) === false)
 		{
 			throw new DatabaseException(lang('Database.failGetIndexData'));
 		}
 
-		if (! $row = $query->getRowArray())
+		if (! $indexes = $query->getResultArray())
 		{
 			return [];
 		}
 
-		$retval = [];
-		foreach (explode("\n", $row['Create Table']) as $line)
+		$keys = [];
+
+		foreach ($indexes as $index)
 		{
-			$line = trim($line);
-			if (strpos($line, 'PRIMARY KEY') === 0)
+			if (empty($keys[$index['Key_name']]))
 			{
-				$obj = new \stdClass();
-				$obj->name = 'PRIMARY KEY';
-				$_fields = explode(',', preg_replace('/^.*\((.+)\).*$/', '$1', $line));
-				$obj->fields = array_map(function($v) {
-					return trim($v, '`');
-				}, $_fields);
-				$obj->type = 'PRIMARY';
-
-				$retval[] = $obj;
-			}
-			elseif (($unique = strpos($line, 'UNIQUE KEY') === 0) || strpos($line, 'KEY') === 0)
-			{
-				if (preg_match('/KEY `([^`]+)` \((.+)\)/', $line, $matches))
+				$keys[$index['Key_name']]       = new \stdClass();
+				$keys[$index['Key_name']]->name = $index['Key_name'];
+
+				if ($index['Key_name'] === 'PRIMARY')
+				{
+					$type = 'PRIMARY';
+				}
+				elseif ($index['Index_type'] === 'FULLTEXT')
 				{
-					$obj = new \stdClass();
-					$obj->name = $matches[1];
-					$obj->fields = array_map(function($v) {
-						return trim($v, '`');
-					}, explode(',', $matches[2]));
-					$obj->type = $unique ? 'UNIQUE' : 'INDEX';
-
-					$retval[] = $obj;
+					$type = 'FULLTEXT';
+				}
+				elseif ($index['Non_unique'])
+				{
+					if ($index['Index_type'] === 'SPATIAL')
+					{
+						$type = 'SPATIAL';
+					}
+					else
+					{
+						$type = 'INDEX';
+					}
 				}
 				else
 				{
-					throw new \LogicException(lang('Database.parseStringFail'));
+					$type = 'UNIQUE';
 				}
+
+				$keys[$index['Key_name']]->type = $type;
 			}
+
+			$keys[$index['Key_name']]->fields[] = $index['Column_name'];
 		}
 
-		return $retval;
+		return $keys;
```

Comment:
The [tests](https://travis-ci.org/natanfelles/CodeIgniter4/jobs/453165355#L611) are failing due to return the $keys indexes with the key name. The test [expect to be keys as integers](https://github.com/bcit-ci/CodeIgniter4/blob/53f17e8ff080b1490cd7df3c57ec2924c858e21b/tests/system/Database/Live/ForgeTest.php#L186).

I think about return the indexes with the key names by default, then update the tests - what do you think? Thanks.

Reply:
With key names will not have issues with ordering, like:

https://github.com/bcit-ci/CodeIgniter4/blob/53f17e8ff080b1490cd7df3c57ec2924c858e21b/tests/system/Database/Live/ForgeTest.php#L184-L207

Reply:
That should be fine.